src/TokenBuyerV2.sol

```diff
@@ -0,0 +1,459 @@
+        uint16 _botDiscountBPs,
+        uint16 _minAdminBotDiscountBPs,
+        uint16 _maxAdminBotDiscountBPs,
+        address _owner,
+        address _admin,
+        address _payer,
+        address _sellToken,
+        address _treasury
+    ) {
+        payer = IPayer(_payer);
+
+        address _paymentToken = address(payer.paymentToken());
+        paymentToken = IERC20Metadata(_paymentToken);
+        paymentTokenUnit = 10**IERC20Metadata(_paymentToken).decimals();
+        priceFeed = _priceFeed;
+
+        baselinePaymentTokenAmount = _baselinePaymentTokenAmount;
+        minAdminBaselinePaymentTokenAmount = _minAdminBaselinePaymentTokenAmount;
+        maxAdminBaselinePaymentTokenAmount = _maxAdminBaselinePaymentTokenAmount;
+
+        if (
+            (_botDiscountBPs > MAX_BPS) ||
+            (_maxAdminBotDiscountBPs > MAX_BPS) ||
+            (_minAdminBotDiscountBPs > _maxAdminBotDiscountBPs)
+        ) {
+            revert InvalidBotDiscountBPs();
+        }
+        botDiscountBPs = _botDiscountBPs;
+        minAdminBotDiscountBPs = _minAdminBotDiscountBPs;
+        maxAdminBotDiscountBPs = _maxAdminBotDiscountBPs;
+
+        _transferOwnership(_owner);
+        admin = _admin;
+        sellToken = IERC20Metadata(_sellToken);
+        sellTokenUnit = 10**IERC20Metadata(_sellToken).decimals();
+        treasury = _treasury;
+    }
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      EXTERNAL TRANSACTIONS
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /// @notice Buy `sellToken` from this contract in exchange for `paymentToken` tokens.
+    /// The price is determined using `priceFeed` plus `botDiscountBPs`
+    /// Immediately invokes `payer` to pay back outstanding debt
+    /// @dev Caps `tokenAmount` by the amount of tokens the contract needs
+    /// @param paymentTokenAmount the amount of ERC20 tokens msg.sender wishes to sell to this contract
+    function swapTokens(uint256 paymentTokenAmount) external nonReentrant whenNotPaused {
+        uint256 amount = Math.min(paymentTokenAmount, paymentTokenAmountNeeded());
+
+        // Cache payer
+        IPayer _payer = payer;
+
+        // Transfer tokens from msg.sender to `payer`
+        paymentToken.safeTransferFrom(msg.sender, address(_payer), amount);
+
+        // Invoke `payer` to pay back outstanding debt
+        _payer.payBackDebt(amount);
+
+        // Send msg.sender STETH
+        uint256 sellTokenAmount = sellTokenAmountPerPaymentTokenAmount(amount);
+        safeSendSellToken(msg.sender, sellTokenAmount);
+
+        emit SwappedTokens(msg.sender, sellTokenAmount, amount);
+    }
+
+    /// @notice Buy sellToken tokens from this contract in exchange for `paymentToken` tokens.
+    /// The price is determined using `priceFeed` plus `botDiscountBPs`
+    /// Immediately invokes `payer` to pay back outstanding debt
+    /// @dev First sends sellToken by calling a callback, and then checks it received payment tokens.
```

Comment:
nit: with ETH we did indeed send value by calling the callback. now we are sending tokens first, then calling the callback to give them a chance to swap. so I would tinker with the wording here